keyring/backends/OS_X.py

```diff
@@ -41,17 +44,40 @@ def set_password(self, service, username, password):
             raise PasswordSetError("Can't store password on keychain: " "{}".format(e))
 
     def get_password(self, service, username):
-        if username is None:
-            username = ''
+        creds={}
+        if (not username):
+            prompt = "find-generic-password"
+            
+            if (service.find(".com")>=0):
+                prompt = "find-internet-password"
+            
+            output = execute(
```

Comment:
This backend seeks not to use the `security` executable to perform its duties, but instead perform its duties with the C APIs... and I definitely want to avoid mixing the two techniques. Can you accomplish the same thing with the APIs (either as implemented or by extending the ctypes adapters)?

Reply:
This is where I'm struggling. I've been scouring Apple's developer documentation for keychain, but of course its all in Objective-C so its only helpful to a certain point. I've been trying two separate approaches, so far neither have been successful.

Approach 1:
Currently you're using `SecKeychainFindGenericPassword` or `SecKeychainFindInternetPassword`. to retrieve the password. If an empty string is passed into the function for the username, the function still returns the password for the specified service, but not the corresponding username. Seeing that the last paramater of this function is an "itemRef" I tried passing this into `SecKeychainItemCopyAttributesAndData` and playing around with this function but the best I was able to do was return the password again.

Approach 2:
I noticed from Apple's documentation that `SecKeychainFindGenericPassword` is a legacy function and that they now recommend the use of `SecItemCopyMatching` which appears to be able to return a reference to all items associated with a keychain entry so in theory this seemed like the most logical approach, however I have thus far been unable to implement this successfully. It starts to get super messy trying when trying to cast things to the expected CF types.

Reply:
Just out of curiosity what advantages to you find in having it implemented using the C APIs as opposed to executing the security shell commands?

Reply:
> what advantages do you find in having it implement using the C APIs?

See the description for #217, which details the motivations.

> I've been trying two separate approaches.

Thanks for detailing the attempts. I'll try to set some time aside in the future to explore the possibilities. The second approach sounds more stable long-term.

Reply:
> ...

I confirmed that `SecKeychainFindGenericPassword` is indeed listed under [Legacy Password Storage](https://developer.apple.com/documentation/security/keychain_services/keychain_items?language=objc#1659107), referencing [Keychain Item Search](https://developer.apple.com/documentation/security/keychain_services/keychain_items?language=objc#1662299) (`SecItemCopyMatching`) as the replacement.

Reply:
I've created #400 to explore Approach 2. I still think Approach 1 has some merit, but I don't know when I'll find the time to flesh it out. My instinct, given the Legacy status, is to focus on Approach 2.